### ISSUE TITLE
setup.py seems to ignore site.cfg's runtime_library_path variable

### ISSUE THREAD

**Issue by nkoep** [numpy/numpy#6435]
Hello,

I'm trying to build a version of numpy that links against a locally built version of openblas. I added the `library_dirs` and `include_dirs` variable to the openblas section in my site.cfg. Running `ldd` confirms that lapack_lite.so links against libopenblas.so. However, for some reason the `runtime_library_dirs` entry (and `rpath` for that matter) in my site.cfg file seems to be ignored. This means that I get an `ImportError` when importing numpy unless I define LD_LIBRARY_PATH before running the interpreter. I fixed the issue by manually setting the rpath via `extra_link_args = -Wl,-rpath,/path/to/openblas/lib`.

Am I misinterpreting the meaning of the `runtime_library_dirs` variable or is this a bug?

**Comment by rgommers** [comment]
``runtime_library_dirs`` was added quite recently: gh-5597. Are you trying to build current master or 1.10.0 (those have it) or an earlier version?

If that's not the issue, I suspect that @zerothi (who wrote this feature) will be able to help you if you provide your full ``site.cfg`` and a link to your build log.

**Comment by zerothi** [comment]
@rgommers @nkoep 
Yes, please provide more information, my successful usage is this snippet:

```
[openblas]
library_dirs = /opt/openblas/0.2.14.2/gnu-5.2.0/lib
include_dirs = /opt/openblas/0.2.14.2/gnu-5.2.0/include
libraries = lapack,openblasp
extra_link_args = -lpthread -lgfortran -lm
runtime_library_dirs = /opt/openblas/0.2.14.2/gnu-5.2.0/lib
```

**Comment by nkoep** [comment]
Sorry for taking so long to reply. I tried to build and link numpy 1.10.1 (pulled from pip) against openblas 0.2.14 (pulled from github). My site.cfg contains
```
[openblas]
library_dirs = /tmp/tmp.eNEY0orve8/src/OpenBLAS-0.2.14/build/lib
include_dirs = /tmp/tmp.eNEY0orve8/src/OpenBLAS-0.2.14/build/include
runtime_library_dirs = /tmp/tmp.eNEY0orve8/src/OpenBLAS-0.2.14/build/lib
```
You can find a complete build log [here](https://gist.github.com/nkoep/7f292b166913840901aa). When I run `ldd` on, e.g., multiarray.so, I get
```
ldd env/lib/python2.7/site-packages/numpy/core/multiarray.so
	linux-vdso.so.1 =>  (0x00007fff55504000)
	libopenblas.so.3 => not found
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f3c40482000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f3c40265000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f3c3feda000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f3c40a99000)
```
Also, if you look at line [1585](https://gist.github.com/nkoep/7f292b166913840901aa#file-gistfile1-txt-L1585) of the log, the correct library search path is added, but no rpath is passed to the linker.

**Comment by zerothi** [comment]
It does not seem to read your site.cfg file at all, where do you have it? Pip installs and site.cfg could be troublesome.
Could you try this (assuming you are using bash):

```
unset LDFLAGS
python setup.py config
```
The output from the config should show whether or not the runtime library directories are read or not, i.e.
```
openblas_info:
/home/nicpa/phd/code/install/.compile/numpy-1.10.1/numpy/distutils/system_info.py:635: 
  FOUND:
    extra_link_args = ['-lpthread -lgfortran -lm']
    define_macros = [('HAVE_CBLAS', None)]
    language = c
    runtime_library_dirs = ['/opt/openblas/0.2.14.2/gnu-5.2.0/lib']
    libraries = ['lapack', 'openblasp']
    library_dirs = ['/opt/openblas/0.2.14.2/gnu-5.2.0/lib']
```

If the config does not spit out something similar to above, could you post the exact commands you execute, from the beginning. :)

**Comment by nkoep** [comment]
I actually wrote a short shell script to build openblas and numpy: https://gist.github.com/nkoep/cfb8946023f40ce930c9

The site.cfg file resides in the root directory of the numpy source tree (right next to site.cfg.example). Running `python setup.py config` produces the following:
```
$ python setup.py config
Running from numpy source directory.
/usr/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'test_suite'
  warnings.warn(msg)
blas_opt_info:
blas_mkl_info:
  libraries mkl,vml,guide not found in ['/usr/local/lib', '/usr/lib', '/usr/lib/x86_64-linux-gnu']
  NOT AVAILABLE

openblas_info:
/tmp/tmp.eNEY0orve8/src/numpy-1.10.1/numpy/distutils/system_info.py:635: UserWarning: Specified path  is invalid.
  warnings.warn('Specified path %s is invalid.' % d)
  FOUND:
    libraries = ['openblas']
    library_dirs = ['/tmp/tmp.eNEY0orve8/src/OpenBLAS-0.2.14/build/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]

  FOUND:
    libraries = ['openblas']
    library_dirs = ['/tmp/tmp.eNEY0orve8/src/OpenBLAS-0.2.14/build/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]

F2PY Version 2
lapack_opt_info:
openblas_lapack_info:
C compiler: cc

creating /tmp/tmpN3l9r_/tmp
creating /tmp/tmpN3l9r_/tmp/tmpN3l9r_
compile options: '-c'
cc: /tmp/tmpN3l9r_/source.c
cc /tmp/tmpN3l9r_/tmp/tmpN3l9r_/source.o -L/tmp/tmp.eNEY0orve8/src/OpenBLAS-0.2.14/build/lib -lopenblas -o /tmp/tmpN3l9r_/a.out
  FOUND:
    libraries = ['openblas']
    library_dirs = ['/tmp/tmp.eNEY0orve8/src/OpenBLAS-0.2.14/build/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]

  FOUND:
    libraries = ['openblas']
    library_dirs = ['/tmp/tmp.eNEY0orve8/src/OpenBLAS-0.2.14/build/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]

/usr/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'define_macros'
  warnings.warn(msg)
running config
```

**Comment by zerothi** [comment]
I remember now why.  
The reason is simply that a runtimepath, and a library path for that matter, is only added to the search path if it passes a library check. However, if one does not set a library to search for it cannot assert the paths given.

So the simple fix is simply to add this to your `[openblas]` block:
```
libraries = openblas
```
and it will work.

@rgommers would you like me to make a PR to change this. It will mean that the default library names are automatically used as libraries when asserting the paths given. 
I refrained from changing this in the PR-5597 as this seemed like a rather drastic change, easy and simple, yet drastic.
I cannot fully determine the implications of this? (it is a 3 line change)

**Comment by nkoep** [comment]
That's so weird. I could have sworn that I tried that before with the same outcome (which is why I eventually removed the `libraries` line from my site.cfg altogether). It seems to work now though. One difference is that openblas is also installed globally on the target system by now, so I can't test what happens if it's not (since I don't have root access).

In any case, I would at least recommend adding a note about it to the site.cfg.example file if you decide not to change the implementation.

Thanks for your help!

**Comment by rgommers** [comment]
@zerothi I think this makes sense to change; don't think that that will break anything. If a user has a section ``[openblas]`` (or another blas) in his ``site.cfg``, then it seems clear that they'd like to use it. A PR would be great.

**Comment by zerothi** [comment]
Ok, I will have a look at it tomorrow :)

--

Kind regards Nick Papior

**Comment by zerothi** [comment]
@rgommers see #7048

**Comment by charris** [comment]
Should be closed by #7048 .
